Refactor fetchRedditPosts to be an API Route aka Serverless Function w/Cache

JeanPec opened the pull request:
https://github.com/erin-fitzpatric/next-aom-gg/issues/96

Refactor fetchRedditPosts in a serverless function w/ Cache

And use public call to fetch the reddit posts

JeanPec committed:
refactor fetchServices and use public url for Reddit posts

erin-fitzpatric commented on src/components/reddit-feed.tsx:
```diff
@@ -1,35 +1,41 @@
-    fetchRedditPosts()
+    fetch(REDDIT_GET_URL)
+      .then((response) => {
+        console.log(response);
```
remove this console log

erin-fitzpatric commented on src/components/reddit-feed.tsx:
```diff
@@ -1,35 +1,41 @@
   const [galleryImages, setGalleryImages] = useState<{ [key: string]: string[] }>({});
 
   function getRedditPosts() {
     // TODO -move this to a serverless function in the /api folder and call it from the client
```
remove this comment

erin-fitzpatric reviewed (changes requested):
![image](https://github.com/user-attachments/assets/db495cf5-86e3-42a9-9f07-c476330595e7)

Looking at the Vercel preview - I'm not seeing the Reddit gallery anymore. This could very likely be the issues with the API I've been facing (and hoping to solve with a cache), but I can't confirm without pulling this down.

At the very least, the skeleton that was there is missing when there is an error response, if you can make sure that gets put back in. I think the skeleton is there because I can see it during loading, but being hidden when there is no response. Ideally there should just always be a response...wish I knew why the Reddit API has been having issues this week. I swear it is rate limit from them or something, but traffic has been down (unless it is a monthly limit or something)

 If it is working for you locally, I can find some time to pull it down.

JeanPec committed:
fix small defects

JeanPec committed:
Merge branch 'erin-fitzpatric:main' into main

JeanPec commented:
<img width="1436" alt="Capture d’écran 2024-09-21 à 17 48 24" src="https://github.com/user-attachments/assets/9de57ad7-3cf5-43bd-a78f-40a2bc3a727a">
Yes it works on my local
I do not really know it is maybe because I switched to the public route, I saw you used an authentifacted route before. I switched to the public one so it would be easier. Maybe it is because of that change.
I am not sure what you want me to do display a placeHodler when I do not get any values ?

JeanPec commented:
retesting it it seems it does not always work I should maybe add a case where I fetch again if I get an error and provide a placeHolder when it fails

erin-fitzpatric commented:
> retesting it it seems it does not always work I should maybe add a case where I fetch again if I get an error and provide a placeHolder when it fails

I have noticed that it has been acting weird in local, but fine in prod. Ya the skelton is ideally just shown when the page is loading, but in the case that it never returns a response, I think it is better to just leave the skelton there rather than hiding the component all together.

JeanPec committed:
Merge branch 'main' into main

JeanPec committed:
add retry

erin-fitzpatric commented on src/components/reddit-feed.tsx:
```diff
@@ -23,15 +25,19 @@ export default function RedditFeed() {
       .catch((error) => {
         console.error("Failed to fetch reddit posts", error);
-      });
+        setTimeout(() => getRedditPosts(), TIMEOUT_RETRY);
```
Does this mean that it will keep retrying in the case that the posts do not load? I think part of the reason they are not loading is likely due to API key limits, but I haven't investigated it throughly.

I'm ok with a retry or 2, but don't want it trying to refecth on a loop. I would rather just show the skelton in the 'loading' state.

erin-fitzpatric replied:
@JeanPec - let me know about this change. Would like to get it out there this weekend.

JeanPec replied:
Okay I will add a count I was not sure if I should do an infinite refresh. Yes we can display the loading if the user really want to get rid of it he can refresh his page

JeanPec replied:
@erin-fitzpatric I just added a limit to the number of retry 👍

erin-fitzpatric commented on src/app/api/reddit/route.ts:
```diff
@@ -0,0 +1,26 @@
+      return new Response(JSON.stringify(mappedPosts), {
+        headers: {
+            "Content-Type": "application/json",
+            "Cache-Control": "public, max-age=43200, stale-while-revalidate=43200",
```
Thanks for adding this cache! Caching this response might resolve the current issues we are seeing with the Reddit fe3ed.

JeanPec replied:
Thanks

JeanPec committed:
Merge branch 'erin-fitzpatric:main' into main

JeanPec committed:
Add max retries

erin-fitzpatric commented:
@JeanPec - thanks for your work on this. We decided to go the server side rendered route with this. https://github.com/erin-fitzpatric/next-aom-gg/pull/144

I am going to go ahead and close this one.
